dags/sentinel-2_data_transfer.py

```diff
@@ -71,100 +94,159 @@ def africa_tile_ids():
+def write_scene(src_key):
+    """
+    Write a file to destination bucket
+    param message: key to write
+    """
+    src_key = src_key
+    s3_hook = S3Hook(aws_conn_id=dag.default_args["africa_conn_id"])
+    s3_hook.copy_object(
+        source_bucket_key=src_key,
+        dest_bucket_key=src_key,
+        source_bucket_name=default_args["src_bucket_name"],
+        dest_bucket_name=default_args["dest_bucket_name"],
+    )
+    return "Done"
 
-def copy_s3_objects(ti, **kwargs):
+
+def start_transfer(message):
     """
-    Copy objects from a s3 bucket to another s3 bucket.
-    :param ti: Task instance
+    Transfer a scene from source to destination bucket
     """
 
-    num_workers = kwargs['num_workers']
-    last_worker_index = num_workers - 1
-    index = kwargs['index']
-    messages = ti.xcom_pull(key='Messages', task_ids='check_for_messages')
-    num_msg_per_worker = len(messages) // num_workers
-    start_index = index * num_msg_per_worker
+    body = json.loads(message.body)
+    metadata = json.loads(body["Message"])
+    tile_id = metadata["id"].split("_")[1]
+
+    s3_hook = S3Hook(aws_conn_id=dag.default_args["africa_conn_id"])
+    s3_hook_oregon = S3Hook(aws_conn_id=dag.default_args["us_conn_id"])
+
+    s3_filepath = (
+        metadata.get("links")[1]["href"]
+        if metadata.get("links")[1].get("rel") == "canonical"
+        else metadata["links"][0]["href"]
+    )
+
+    # Check file exists
+    key = extract_src_key(s3_filepath)
+    key_exists = s3_hook_oregon.check_for_key(
+        key, bucket_name=default_args["src_bucket_name"]
+    )
+    if not key_exists:
+        raise ValueError(
+            f"{key} does not exist in the {default_args['src_bucket_name']} bucket"
+        )
+
+    urls = [s3_filepath]
+
+    print(f"Copying {Path(s3_filepath).parent}")
+    # Add URL of .tif files
+    urls.extend(
+        [v["href"] for k, v in metadata["assets"].items() if "geotiff" in v["type"]]
+    )
+
+    # Check that all bands and STAC exist
+    if len(urls) != 18:
+        raise ValueError(
+            f"There are less than 18 files in {metadata.get('id')} scene, failing"
+        )
+
+    src_keys = []
+    for src_url in urls:
+        src_key = extract_src_key(src_url)
+        key_exists = s3_hook_oregon.check_for_key(
+            key, bucket_name=default_args["src_bucket_name"]
+        )
+        src_keys.append(src_key)
+
+        if not key_exists:
+            raise ValueError(
+                f"{key} does not exist in the {default_args['src_bucket_name']} bucket"
+            )
+
+    with ThreadPoolExecutor(max_workers=5) as executor:
+        task = {executor.submit(write_scene, key): key for key in src_keys}
+        for future in as_completed(task):
+            scene_to_copy = task[future]
+            try:
+                result = future.result()
+            except Exception as exc:
+                raise ValueError(f"{scene_to_copy} failed to copy")
```

Comment:
I don't think this is a good idea. I think we need better error handling when one scene fails to copy. How does this work when one out of lots did copy?

Reply:
Files that are copied will remain in the destination bucket.  Since the message fails.  Since we know all the files exist, the next try should cover this.

What do you mean by better error handling, like separately reporting these types of failures?